pkg/controller/backup_sync_controller.go

```diff
@@ -326,73 +287,149 @@ func (c *backupSyncController) run() {
 						log.WithError(errors.WithStack(err)).Errorf("Error syncing volumesnapshotcontent %s into cluster", snapCont.Name)
 						continue
 					default:
-						log.Infof("Created CSI volumesnapshotcontent %s", created.Name)
+						log.Infof("Created CSI volumesnapshotcontent %s", snapCont.Name)
 					}
 				}
 			}
 		}
 
-		c.deleteOrphanedBackups(location.Name, backupStoreBackups, log)
+		b.deleteOrphanedBackups(ctx, location.Name, backupStoreBackups, log)
 
 		// update the location's last-synced time field
 		statusPatch := client.MergeFrom(location.DeepCopy())
 		location.Status.LastSyncedTime = &metav1.Time{Time: time.Now().UTC()}
-		if err := c.kbClient.Patch(context.Background(), &location, statusPatch); err != nil {
+		if err := b.client.Patch(ctx, &location, statusPatch); err != nil {
 			log.WithError(errors.WithStack(err)).Error("Error patching backup location's last-synced time")
 			continue
 		}
 	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager is used to setup controller and its watching sources.
+func (b *backupSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	backupSyncSource := kube.NewPeriodicalEnqueueSource(
+		b.logger,
+		mgr.GetClient(),
+		&velerov1api.BackupStorageLocationList{},
+		backupSyncReconcilePeriod,
+		// Only enqueue the first BSL
+		func(object client.Object) bool {
```

Comment:
How about that we change the Reconciler to only handle one BSL at a time, so that we can enqueue all items rather than just the first one. The current code in this PR seems a little weird, the `PeriodicalEnqueueSource` is just used as a signal, and the `List` is called n+1 times(n is the count of BSL).

For the BSL ordering, how about we define an OrderFunc for the `PeriodicalEnqueueSource` so that the caller can customize the order of the items.

Reply:
Basically I agree with your suggestion.
Per my understanding, since adding a new orderFunc, we need a default orderFunc to return input list as output for other controller using `PeriodicalEnqueueSource`.

Reply:
How about that we define an `Option` struct for the `PeriodicalEnqueueSource`, the `Option` is extensible easily without modifying the caller code when adding new features to it. Currently, it contains only `FilterFunc` and `OrderFunc`